fix : making on-screen/soft keyboard dismiss on background tap

gururani-abhishek opened the pull request:
Fixes #1499 

To enhance the user flow, added the ability to dismiss the soft keyboard or on-screen keyboard by tapping on the background or anywhere other than an EditText.

1. dismissSoftKeyboardOnBkgTap() takes View/ViewGroup as an input and then individually checks all its child Views/ViewGroups to find a View that doesn't include EditText.

2. when such a view is found and a tap is made on such a view, it's passed to hideSoftKeyboard() Utils function that by using Input Method Manager hides the soft keyboard.

3. tested on pixel 2(api 33), the fix works smoothly.



https://user-images.githubusercontent.com/59947871/226180530-0580d102-6e02-41d0-b924-9cf536055348.mp4




Please Add Screenshots If there are any UI changes.

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Apply the `AndroidStyle.xml` style template to your code in Android Studio.

- [x] Run the unit tests with `./gradlew check` to make sure you didn't break anything

- [x] If you have multiple commits please combine them into one commit by squashing them.

gururani-abhishek committed:
fix #1499 : making on-screen/soft keyboard dismiss on background tap

To enhance the user flow, added the ability to dismiss the soft keyboard or on-screen keyboard by
tapping on the background or anywhere other than an EditText.

1. dismissSoftKeyboardOnBkgTap() takes View/ViewGroup as an input and then individually checks
all its child Views/ViewGroups to find a View that doesn't include EditText.

2. when such a view is found and a tap is made on such a view, it's passed to hideSoftKeyboard() Utils function
that by using Input Method Manager hides the soft keyboard.

3. tested on pixel 2(api 33), the fix works smoothly.

PratyushSingh07 commented:
there is already PR #1942 for this i guess

gururani-abhishek commented:
> there is already PR #1942 for this i guess

Oh, seems like they created a whole new issue for an existing one.

PratyushSingh07 commented:
i reckon so @gururani-abhishek , there are a total of 4 PRs for the same issue . This maybe because some of them are really old PRs ranging way back to 2020

gururani-abhishek commented:
> there is already PR #1942 for this I guess

with dispatchTouchEvent() solution if you tap a different location of the EditText to move the input cursor, it hides the keyboard, and the keyboard pops up again by the system. There are some UX related issues that comes along with it.

gururani-abhishek commented:
> i reckon so @gururani-abhishek , there are a total of 4 PRs for the same issue . This maybe because some of them are really old PRs ranging way back to 2020

Yeah, I myself took great help from the pre-existing ones. But due to changes in the xml, and migration of LoginActivity from Java to Kotlin, they've become obsolete considering our codebase.

jawidMuhammadi commented on app/src/main/java/org/mifos/mobile/ui/activities/LoginActivity.kt:
```diff
@@ -57,6 +56,27 @@ class LoginActivity : BaseActivity(), LoginView {
+
+    private fun dismissSoftKeyboardOnBkgTap(view: View) {
+
+        // If this individual view doesn't include an EditText
```
Remove this comment, please.

gururani-abhishek replied:
done :)

jawidMuhammadi commented on app/src/main/java/org/mifos/mobile/ui/activities/LoginActivity.kt:
```diff
@@ -57,6 +56,27 @@ class LoginActivity : BaseActivity(), LoginView {
+        }
+
+        //If it's a ViewGroup(or layout container), iterate over all its child views and
+        // recursively call dismissSoftKeyboardOnBkgTap() for each individual view.
```
This comment too

gururani-abhishek replied:
done :)

jawidMuhammadi commented on app/src/main/java/org/mifos/mobile/utils/Utils.kt:
```diff
@@ -77,4 +81,11 @@ object Utils {
         return builder.toString()
     }
+
+    fun hideSoftKeyboard(activity: Activity) {
```
I believe we already have this method implemented in the BaseActivity, use that one, please.

gururani-abhishek replied:
Ops that's totally my fault, I'll be extra considerate of existing code now onwards. Thanks @jawidMuhammadi. I've made the required changes.

gururani-abhishek committed:
fix #1499 : : making on-screen/soft keyboard dismiss on background tap

To enhance the user flow, added the ability to dismiss the soft keyboard or on-screen keyboard by
tapping on the background or anywhere other than an EditText.

1. dismissSoftKeyboardOnBkgTap() takes View/ViewGroup as an input and then individually checks
all its child Views/ViewGroups to find a View that doesn't include EditText.

2. when such a view is found and a tap is made on such a view, it's passed to hideKeyboard() that by using
Input Method Manager hides the soft keyboard.

3. tested on pixel 2(api 33), the fix works smoothly.

jawidMuhammadi commented on app/src/main/java/org/mifos/mobile/utils/Utils.kt:
```diff
@@ -9,9 +10,12 @@ import android.graphics.drawable.LayerDrawable
 import android.net.Uri
 import android.util.Log
 import android.view.Menu
+import android.view.inputmethod.InputMethodManager
```
Are you sure this import and the others which are added by you are being used?

gururani-abhishek replied:
No they aren't, I've removed them now.

jawidMuhammadi commented on app/src/main/java/org/mifos/mobile/ui/activities/LoginActivity.kt:
```diff
@@ -57,6 +55,24 @@ class LoginActivity : BaseActivity(), LoginView {
+    }
+
+    private fun dismissSoftKeyboardOnBkgTap(view: View) {
+
```
Remove this space, please

gururani-abhishek replied:
you mean line 62, right?

jawidMuhammadi replied:
Yes, that is right

jawidMuhammadi commented on app/src/main/res/layout/activity_login.xml:
```diff
@@ -3,6 +3,7 @@
     xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:app="http://schemas.android.com/apk/res-auto"
     android:fitsSystemWindows="true"
+    android:id="@+id/nsv_bkg"
```
what does the `nsv_bkg` stand for? Could you please give it's full name?

gururani-abhishek replied:
It's NestedScrollView which is the background. I found places where if a LinearLayout is used as a login tray its id is given as "ll_login" and hence I gave it an id "nsv_bkg", should I rename it?

jawidMuhammadi replied:
Yes, something to: `nsv_background`

gururani-abhishek replied:
done!!

jawidMuhammadi commented on app/src/main/java/org/mifos/mobile/ui/activities/LoginActivity.kt:
```diff
@@ -57,6 +55,24 @@ class LoginActivity : BaseActivity(), LoginView {
         setContentView(R.layout.activity_login)
         ButterKnife.bind(this)
         loginPresenter?.attachView(this)
+        dismissSoftKeyboardOnBkgTap(findViewById(R.id.nsv_bkg))
```
hmmm... as you you are binding the view with ButterKnife, you could access the view directly by its id, I think there is no need for `findViewById`, agree?

gururani-abhishek replied:
Yes I agree. Made the changes.

gururani-abhishek committed:
fix #1499 : making on-screen/soft keyboard dismiss on background tap

To enhance the user flow, added the ability to dismiss the soft keyboard or on-screen keyboard by
tapping on the background or anywhere other than an EditText.

1. dismissSoftKeyboardOnBkgTap() takes View/ViewGroup as an input and then individually checks
all its child Views/ViewGroups to find a View that doesn't include EditText.

2. when such a view is found and a tap is made on such a view, it's passed to hideKeyboard() that by using
Input Method Manager hides the soft keyboard.

3. tested on pixel 2(api 33), the fix works smoothly.# Please enter the commit message for your changes. Lines starting

jawidMuhammadi reviewed (approved):
LGTM, great Job!

gururani-abhishek commented:
> LGTM, great Job!

Thanks @jawidMuhammadi, learned a lot while working on this. How about if I extend this feature to other screens too? Like the Register Screen, and every screen where the soft keyboard is used.
